### PR TITLE
Add OneBit mode to HARDWARE menu.

### PR DESCRIPTION
**Pull request by hydra**
OneBit mode can help if your radio has a slow inverter circuit by changing the UART sampling behaviour.

* It was found that the X9D+ and X7 (maybe others) have a slow inverter that messes up reception of fast (400,000baud) UART signals.  On the scope the rise time of the signal is very long.
* See RM0033 24.3.3 Receiver / Selecting the proper oversampling method.  Page 63.

Also note that reducing the pull-up resistor from 10k to 1k at R82 (X9D+) / R406 (X7) will also help with the rise times.  However on the X7 changing the resistor is not enough, OneBit is needed too.

IMG_20210520_213949

Inverter output measure at MCU with standard 10k pull-up resistor:

Inverter output measure at MCU with standard 1k pull-up resistor:

Here's how using OneBit changes the sampling.
onebit

When using OneBit, the signal is sampled later, slow leading edges are ignored.

### CODE REVIEW

**Comment by raphaelcoeffic** [comment]
@hydra the new field in RadioData struct will have to be moved to the end of the structure to avoid storage format conversion. Another option would be to find some spare bits to convert. Unfortunately, I cannot do it for you in this PR unless you invite me to your fork.

**Comment by hydra** [comment]
yeah, I noted in the chat that adding a field in the middle is probably bad. I'll move it to the end.

**Comment by hydra** [comment]
Tested on X9D+ and X7 with a Ghost TX module flashed with ELRS.

Before `OneBit` mode is enabled the ELRS LUA script just displays `Connecting...`, once `OneBit` is enabled the LUA script connects to the Ghost TX module just fine!

**Comment by hydra** [comment]
Binaries for testing.

[firmware-x7-2.3.12-1a7f6c294.zip](https://github.com/EdgeTX/edgetx/files/6518549/firmware-x7-2.3.12-1a7f6c294.zip)
[firmware-x9d+-2.3.12-1a7f6c294.zip](https://github.com/EdgeTX/edgetx/files/6518550/firmware-x9d%2B-2.3.12-1a7f6c294.zip)

File: radio/src/gui/common/stdlcd/radio_hardware.cpp [hunk 1]
```diff
@@ -321,6 +325,20 @@ void onFactoryResetConfirm(const char * result)
...
+  }
+  pauseMixerCalculations();
+  pausePulses();
+  EXTERNAL_MODULE_OFF();
+  RTOS_WAIT_MS(20); // 20ms so that the pulses interrupt will reinit the frame rate
+  telemetryProtocol = 255; // force telemetry port + module reinitialization
+  EXTERNAL_MODULE_ON();
+  resumePulses();
+  resumeMixerCalculations();
+}
```
Comment (raphaelcoeffic) [inline radio/src/gui/common/stdlcd/radio_hardware.cpp:341]: Can I buy an empty line please 😇

**Review by raphaelcoeffic — APPROVED** [review]

File: radio/src/gui/common/stdlcd/radio_hardware.cpp (hunk 1 above)
Comment (raphaelcoeffic) [reply radio/src/gui/common/stdlcd/radio_hardware.cpp:341]: (after the function)

Comment (hydra) [reply radio/src/gui/common/stdlcd/radio_hardware.cpp:341]: yes, yes you can :D

**Comment by brainbubblersbest** [comment]
X10 and x12 Radios have the same issue with crsf Telemetry. If that solution would be applyable to that Radios then the Inverter Mod would be total obsolete.

**Comment by hydra** [comment]
> X10 and x12 Radios have the same issue with crsf Telemetry. If that solution would be applyable to that Radios then the Inverter Mod would be total obsolete.
> 
> [Radios Adresses with Inverter Mod, Blog from Philip Seidel](https://blog.seidel-philipp.de/fixed-inverter-mod-for-tbs-crossfire-and-frsky-qx7/)

Agreed!

**Comment by brainbubblersbest** [comment]
I cannot Do that with the Lack of Talent i have. But it would be Nice when you bring that Feature to more Radio Models.
Ive done already the inverter Mod. But a lot users wont solder on their Hardware for using elrs and crossfire.

**Comment by 1finsurfer** [comment]
I've already flashed EdgeTX and changed the Sample Mode to "OneBit" as pictured above.  I haven't done the resistor mod yet, but should I also set the Serial Port to "LUA" too?

**Comment by samuraitheta** [comment]
I am having trouble finding R82 on the X9D+ Do you have a picture or diagram? Thanks!

**Comment by pfeerick** [comment]
This looks to be it on my X9D+ mainboard... yes, I have one lying around on my desk at the moment 😆 
image

**Comment by oermenz** [comment]
Are they any downsides when using OneBit compared to the hardware mod? Is there such a thing as free lunch? I will be switching to ELRS and have no trouble doing the mod if it provides an advantage over this (excellent?) software work around.

**Comment by hydra** [comment]
> Are they any downsides when using OneBit compared to the hardware mod? Is there such a thing as free lunch? I will be switching to ELRS and have no trouble doing the mod if it provides an advantage over this (excellent?) software work around.

In this case, no, since it works better than without it on!  The OneBit mode on the CPUs have their uses vs the default sampling, refer to the ST documentation as it's a complicated topic.

**Comment by oermenz** [comment]
Yes, I noticed it's complicated and tbh it goes a bit (ha!) over my head. What I get from it is that the clock deviation tolerance is decreased which in turn could increase the error percentage. But all in all it seems that is negligible since the higher rates will also compensate for the minor chance of increased errors. Many thanks for this solution and effort!

**Comment by abimaputra78** [comment]
Pls somebody advice, I had annoying telemetry warning on radio master tx16. Can solve until now.

**Comment by pfeerick** [comment]
How does telemetry warning on RM TX16S relate to OneBit on B&W radios? What do you mean by "annoying telemetry warning"? When does it happen? What RF link (RF protocol/transmitter module/receiver) are you using?

**Comment by kimman1** [comment]
The error is still present when modded R82 to 1K and edgeTX. 
Hardware:
X9D+ on latest EdgeTX (master branch)
DIY ERLS 2.4Ghz 
Behavior: 
400k Baudrate => still good, Lua script work ok
>400k Baudrate => Lua script just show connecting

**Comment by brainbubblersbest** [comment]
> How does telemetry warning on RM TX16S relate to OneBit on B&W radios? What do you mean by "annoying telemetry warning"? When does it happen? What RF link (RF protocol/transmitter module/receiver) are you using?
> 
> Edit: I see you have posted over in #1188 also. Please refrain from asking the same question in multiple places. It would be better to continue the conversation over there rather than commenting further in this unrelated PR.

Just to create a relation to that question. 
Do the horus processors Support also one bit sampling Mode? Because some Horus Radios would clearly benefit from this Feature. 😉

**Comment by 1finsurfer** [comment]
> The error is still present when modded R82 to 1K and edgeTX. Hardware: X9D+ on latest EdgeTX (master branch) and be modded R82 to 1K ohm DIY ERLS 2.4Ghz Behavior: 400k Baudrate => still good, Lua script work ok Larger than 400k Baudrate => Lua script just show connecting Do we have another solution for that ? . Just like mod an inverter ?

Same here, higher than 400k baud rate causes ELRS Lua to be stuck at "Loading...", once power cycled after setting baud rate back to 400k Lua script works.

**Comment by pdxtigre** [comment]
So what's the verdict?
I got an X9D+ ACCST with EDGE TX 2.6 as well.  Seems like the HW mod is not required/necessary?

**Comment by haybeseret** [comment]
> 

HI -I have flash today 2.6 - and the option is there (i.e onbit mode integrated already)

**Comment by haybeseret** [comment]
> > 
> 
> HI -I have flash today 2.6 - and the option is there (i.e onbit mode integrated already)



> So what's the verdict? I got an X9D+ ACCST with EDGE TX 2.6 as well. Seems like the HW mod is not required/necessary?

HI -I have flash today 2.6 - and the option is there (i.e onbit mode integrated already)

**Comment by pfeerick** [comment]
> HI -I have flash today 2.6 - and the option is there (i.e onbit mode integrated already)

Um, was enabled for X9D etc from 2.4 onwards... ;) He's asking about if the HW mod is still required/necessary;)

OneBit doesn't _entirely_ negate the need for the hardware mod, but will be enough in a lot of instances. And as indicated in the very first post, in some instances BOTH the hardware mod and OneBit will be needed.

**Comment by hosein-gh** [comment]
does changing 10k resistor to 1k will affect internal module performance ? or will it make any issue with it ?
i'm using r-xsr on my indoor quadcopters and i don't want to change receiver on them.

**Comment by hosein-gh** [comment]
@hydra @pfeerick does changing 10k resistor to 1k will affect internal module performance ? or will it make any issue with it ?
i'm using r-xsr on my indoor quadcopters and i don't want to change receiver on them.

**Comment by hydra** [comment]
@hosein-gh feel free to check it on a scope, but in general you should use the highest resistor value that allows it to work to limit current and increase TX battery life.  Start high (10k) then try lower values.

**Comment by PScott1** [comment]
I did the resister mod on my X9D with a 10k and it seems to be working fine with OpenTx.

**Comment by L1cardo** [comment]
Is OneBit necessary for RadioMaster TX12, because I am seeing an option there and OneBit is something like a X9D problem, but what is strange is that there is no this option on a colored screen radio(TX16S), so I am really confused, what radios have this OneBit option?
Or can we make this option appears based on radio hardware?

**Comment by ChasedMonkey** [comment]
Hi this option is missing from the horus build for color screen radios. I can't find it on my x10s (which needs it) or on my tx16s.

**Comment by venom-code** [comment]
Taranis X9D+ se 2019, are these modifications needed for this tx to run ELRS, I ask because it works without trouble on Arduplane, tbs crossfire (lua works flawless) and running yappu telemetry script?

**Comment by pfeerick** [comment]
Generally *if it works* means you don't need it (OneBit). It is another thing to try *when it doesn't work*, before you move on to more drastic measures like a hardware mod.

**Comment by venom-code** [comment]
> Generally _if it works_ means you don't need it (OneBit). It is another thing to try _when it doesn't work_, before you move on to more drastic measures like a hardware mod.

thanks for the reply

**Comment by venom-code** [comment]
just migrated from opentx and installed ETX 2.7.1 on a Qx7-accst, everything went smoothly, the Onebit sampling was chosen underneath is an option of internal jst, what is it for? turned off while using an external module? and it seems the baud rate option is gone from the Hardware page on the tx and is now supposed to be set in the Radio setup page ?

**Comment by Willy163** [comment]
**X9D Plus 2019: NO R82**

Hello everybody.
I own the X9D with new motherboard and there is no R82.
So I did some reverse engineering following tracks from the microcontroller pin 87 (USART2_RX).
The resistor that has to be replaced: **R53** 10kOhm to 1kOhm.

That's my 2 cents.

Bye,
Willy.
R53

**Comment by flybabo** [comment]
I don't think you need to replace the pull-up resistor of the X9D+2019.
Old X9D/X9D+ uses an NPN transistor to invert the signal and it has lousy switching characteristics.
The new Taranis, X9D+2019, apparently uses a tristate inverter (U3, C40R=SN74LVC2G240DCUR) which is capable of switching the signal easily over 10MHz.  The purpose of 10K pull-up resistor is to improve the noise margin - not a load resistor.
That's my $0.02 :)

**Comment by NorgeSquad** [comment]
AKA. If it don't make so you can run 921k baud. its not worth trying

**Comment by pfeerick** [comment]
If you want faster packet rates than 333hz Full Res, sure. But you also shouldn't be running hardware that old and expecting it to work with later generation high speed protocols. :shrug:

**Comment by NorgeSquad** [comment]
Yeah im aware. But Would be cool if it did

**Comment by NorgeSquad** [comment]
But I will give it a shot. I'll let you guys know if I can run 921k baud or not

**Comment by joshbasquez** [comment]
> But I will give it a shot. I'll let you guys know if I can run 921k baud or not

so how fast did you get it working in x9d?

**Comment by NorgeSquad** [comment]
> > But I will give it a shot. I'll let you guys know if I can run 921k baud or not
> 
> so how fast did you get it working in x9d?

I only made mine from before the 2019 model run at max 400K but the 2019 edition can do 921k out the box

**Comment by LukeOrgana** [comment]
Im a bit late to the game here, but does anyone know if the TBS Crossfire inverter mod for the QX7 would work on a pre-2019 X9D+? Is this even relevant for ELRS? If so, is the X9D+ motherboard similar enough to the QX7 that the same instructions could be followed? Or is it completely different? Or is the resistor change + one bit mode all that is required?

**Comment by pfeerick** [comment]
AFAIK the best you could get on the X9D+ was 400k, with OneBit. I never did the resistor mod myself - never saw a need as 400k was perfectly stable on mine with just the OneBit mode enabled. Since the ELRS notes only mention the resistor mod, either no-one has tried to do it (which I highly doubt, unless it simply isn't practical) or it just isn't possible. 

Some of the notes there are out of date quite a bit now though, as I know for a fact that 500Hz is now possible @ 400k, and you can also run the D500 and F500 modes also on X9D+. So, unless you want the really high speeds, at 400k baud rate, the X9D+ can still run most of the ELRS packet rates. 

https://www.expresslrs.org/hardware/x9d-troubleshooting/#troubleshoot

**Comment by NorgeSquad** [comment]
My Old non 2019 X9D+ Actually runs on 500Hz Packet Rate. But The Radio itself don't function cuz it uses all the memory to run the 500Hz. But Best stable is 333Hz and Radio don't act up or Freeze like it does on 500Hz

**Comment by pfeerick** [comment]
Yeah, 500Hz will be pushing it on that hardware... I only used to use 250Hz on it. I almost never use anything higher than 333 as I really don't see the point, even on current gen radios. 🤷

**Pull request merged**